fl4health/clients/fenda_client.py

```diff
@@ -1,12 +1,229 @@
+        assert isinstance(self.model, FendaModel)
+        if self.contrastive_loss_weight or self.perFCL_loss_weights:
+            self.old_local_module = self.clone_and_freeze_model(self.model.local_module)
+            self.old_global_module = self.clone_and_freeze_model(self.model.global_module)
+
+        # Set the parameters of the model
+        super().set_parameters(parameters, config)
+
+        # Save the parameters of the global model
+        if self.perFCL_loss_weights:
+            self.aggregated_global_module = self.clone_and_freeze_model(self.model.global_module)
+
+        return
+
+    def get_cosine_similarity_loss(self, local_features: torch.Tensor, global_features: torch.Tensor) -> torch.Tensor:
+        """
+        Cosine similarity loss aims to minimize the similarity among current local features and current global
+        features of fenda model.
+        """
+        assert len(local_features) == len(global_features)
+        return torch.abs(self.cos_sim(local_features, global_features)).mean()
+
+    def get_contrastive_loss(
+        self, local_features: torch.Tensor, old_local_features: torch.Tensor, global_features: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Contrastive loss aims to enhance the similarity between the current local features and old local features
+        as positive pairs while reducing the similarity between the current local features and current global
+        features as negative pairs.
+        """
+        assert isinstance(self.temperature, float)
+        assert len(local_features) == len(old_local_features)
+        posi = self.cos_sim(local_features, old_local_features)
+        logits = posi.reshape(-1, 1)
+        assert len(local_features) == len(global_features)
+        nega = self.cos_sim(local_features, global_features)
+        logits = torch.cat((logits, nega.reshape(-1, 1)), dim=1)
+        logits /= self.temperature
+        labels = torch.zeros(local_features.size(0)).to(self.device).long()
+
+        return self.ce_criterion(logits, labels)
+
+    def get_perFCL_loss(
+        self,
+        local_features: torch.Tensor,
+        old_local_features: torch.Tensor,
+        global_features: torch.Tensor,
+        old_global_features: torch.Tensor,
+        aggregated_global_features: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        PerFCL loss consists of two contrastive losses.
+        First one airms to enhance the similarity between the current global features and aggregated global feature
+        as positive pairs while reducing the similarity between the current global features and old global
+        features as negative pairs.
+        Second one airms to enhance the similarity between the current local features and old local feature
+        as positive pairs while reducing the similarity between the current local features and aggregated global
+        features as negative pairs.
```

Comment:
It might be helpful, for a reader of the code to connect each of the inputs to this loss to the variables in the paper. Something like 
```
local features (z_p)
old_local_features (\hat{z_p})
global_features (z_s)
old_global_features (\hat{z_s})
aggregated_global_features (z_g)
```